main.nf

```diff
@@ -124,76 +132,138 @@ process BWA_ALIGN {
  * Convert the format of the alignment to sorted BAM.
  */
 process SAMTOOLS_SORT {
-  tag{"SAMTOOLS_SORT ${sample_id}"}
-  label 'process_low'
+    tag{"${sample_id}"}
+    label 'process_low'
+    conda 'samtools'
 
-  publishDir("${params.outdir}/bam_align", mode: 'copy')
+    publishDir("${params.outdir}/bam_align", mode: 'copy')
 
-  input:
-  tuple val( sample_id ), path( bam )
+    input:
+    tuple val( sample_id ), path( bam )
 
-  output:
-  tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam" ), emit: sorted_bam
+    output:
+    tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam" ), emit: sorted_bam
 
-  script:
-  """
-  samtools sort -o "${sample_id}.aligned.sorted.bam" ${bam}
-  """
+    script:
+    """
+    samtools sort -o "${sample_id}.aligned.sorted.bam" ${bam}
+    """
 }
 
 /*
  * Index the BAM file for visualization purpose
  */
 process SAMTOOLS_INDEX {
-    // TODO
+    tag{"${sample_id}"}
+    label 'process_low'
+    conda 'samtools'
+
+    publishDir("${params.outdir}/bam_align", mode: 'copy')
+
+    input:
+    tuple val( sample_id ), path( bam )
+
+    output:
+    tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam.bai" ), emit: sorted_bam_index
+
+    script:
+    """
+    samtools index ${bam}
+    """
 }
 
 /*
  * Calculate the read coverage of positions in the genome.
- */
+*/
 process BCFTOOLS_MPILEUP {
-    // TODO
+    tag{"${sample_id}"}
+    label 'process_high'
+    conda "bioconda::bcftools=1.17"
+
+    publishDir("${params.outdir}/bcftools_mpileup", mode: 'copy')
+
+    input:
+    tuple val( sample_id ), path( bam )
+
+    output:
+    tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam.bcf" ), emit: bcf
+
+    script:
+    """
+    bcftools mpileup -O b -o ${sample_id}.aligned.sorted.bam.bcf -f ${params.genome} ${bam}
```

Comment:
```suggestion
    bcftools mpileup -O b -o ${sample_id}.aligned.sorted.bam.bcf -f ${params.genome} ${bam}
```

This throws an error if you use a URL because it can't find the `fai` associated with the fasta file. Check out the change in the ganymede branch that I made if you're interested.